### PR TITLE
metis/parmetis/oce/boost: correct install_name on Darwin via a global tool

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

**Comment by davydden** [comment]
i see that this is being discussed in https://github.com/LLNL/spack/pull/439 but that branch has conflicts and @eschnett argued that there are not many packages that need this hack.

**Comment by davydden** [comment]
@tgamblin maybe we should pull
```
            fs = glob.glob(join_path(prefix.lib,"*.dylib"))
            install_name_tool = which('install_name_tool')
            for f in fs:
                install_name_tool('-id',f,f)

```
into a function somewhere to be called from each package like
```
if (sys.platform == 'darwin'):
    call_me_here()
```
or
```
if (sys.platform == 'darwin') and ('+shared' in spec):
    call_me_here()
```
?

**Comment by davydden** [comment]
it is getting messy for packages with multiple libs (i.e. `boost`, `oce` and `trilinos`). Here is a relevant script to fix the paths: https://github.com/luca-heltai/d2-parallel-package/blob/master/relocate_libraries.py

**Comment by davydden** [comment]
I am working on a general script to do that, almost done...

**Comment by davydden** [comment]
if anyone will come-up with this thread, here is a stand-along python script to do the job:
```
libs = glob.glob(os.path.join(basedir,"*.dylib"))
for lib in libs:
    print('Processing', lib)
    run(["install_name_tool",'-id',lib, lib])
    long_deps = run(["otool",'-L',lib]).split('\n')
    deps = [dep.partition(' ')[0][1::] for dep in long_deps[2:-1]]
    for dep in deps:
        print('--', dep)
        for loc in libs:
            if dep == os.path.basename(loc): # dep.find(os.path.basename(loc)) != -1 and dep != loc:
                print('-- Change ',dep,' to ', loc, ' in ',lib)
                run(["install_name_tool",'-change',dep, loc,lib])
                break
```
i will now adopt it inside packages...

**Comment by davydden** [comment]
i strange to me error on Travis
```
   from subprocess import check_output as run
ImportError: cannot import name check_output
```
i suppose @tgamblin would suggest relocate the function somewhere else anyway. But all in all, this PR is ready for review.

**Comment by davydden** [comment]
for a record, this worked for `trilinos`, which has dozens of `.dylib` files :smile:

**Comment by davydden** [comment]
@tgamblin ping  :eyes:

File: lib/spack/llnl/util/filesystem.py [hunk 1]
```diff
@@ -392,3 +394,29 @@ def remove_linked_tree(path):
...
         else:
             shutil.rmtree(path, True)
+
+def fix_darwin_install_name(path):
+    """
+    Fix install name of dynamic libraries on Darwin to have full path.
+    There are two parts of this task:
+    (i) use install_name('-id',...) to change install name of a single lib;
+    (ii) use install_name('-change',...) to change the cross linking between libs.
+    The script assumes that all libraries are in one folder and currently won't
```
Comment (tgamblin) [inline lib/spack/llnl/util/filesystem.py]: script -> function

**Comment by tgamblin** [comment]
@davydden: I like it.  It's easy to see where the darwin stuff is, so we can go fix up the packages later when there's a cleaner way.

I still wonder whether adding something to `lib/spack/env/cc` to set `RPATH` with a proper OS X `@rpath` prefix would work better, but I don't have time to investigate that right now.  Adding handling to `cc` would, I think, get more automatic coverage (so you don't have to call this many times for packages that have libraries in many locations).  But this should at least get OS X support running soon, and it's good for me because I'm honestly not that familiar with OS X support so I can see what's necessary.

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -38,6 +39,7 @@
 import getpass
 from contextlib import contextmanager, closing
 from tempfile import NamedTemporaryFile
+from subprocess import check_output as run
```
Comment (tgamblin) [inline lib/spack/llnl/util/filesystem.py]: Travis is complaining about this, I think.  Can you look at the broken travis build?

**Comment by davydden** [comment]
@tgamblin i could play around with rpath as well, but it's (i) more work (ii) one does not need it everywhere (in each package). So i have preferences to the current approach, at least for now.

Any idea why travis is unhappy?

File: lib/spack/llnl/util/filesystem.py (hunk 2 above)
Comment (davydden) [inline lib/spack/llnl/util/filesystem.py]: i don't quite understand what is wrong:
```
 File "/home/travis/build/LLNL/spack/lib/spack/llnl/util/filesystem.py", line 42, in <module>
    from subprocess import check_output as run
ImportError: cannot import name check_output
```

Comment (davydden) [inline lib/spack/llnl/util/filesystem.py]: it seems this function is not available in `2.6`
```
[1;31m==>[0m Error: Spack must run on Python version 2.6
[1;31m==>[0m Error: These files require version 2.7:
F.
======================================================================
FAIL: test_core_module_compatibility (spack.test.python_version.PythonVersionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/LLNL/spack/lib/spack/spack/test/python_version.py", line 100, in test_core_module_compatibility
    self.check_python_versions(*self.pyfiles(spack.lib_path))
  File "/home/travis/build/LLNL/spack/lib/spack/spack/test/python_version.py", line 96, in check_python_versions
    self.assertTrue(len(all_issues) == 0)
AssertionError: False is not true
-------------------- >> begin captured stdout << ---------------------
File                                         Reason
------------------------------------------   --------------------
spack/lib/spack/llnl/util/filesystem.py:42   subprocess.check_output

--------------------- >> end captured stdout << ----------------------
```

**Comment by davydden** [comment]
will squash in a moment, checking that i did not break it along the way...

**Comment by davydden** [comment]
@tgamblin all seems to be good.

**Pull request merged**